### PR TITLE
Don't suggest alias fragments with leading number

### PR DESCRIPTION
**Pull request by pdbrown**
Hey there! By pure coincidence, I happened to be investigating a possible bug in `cljr-slash` and `refactor-nrepl.ns.suggest-aliases/suggested-aliases` just as you posted this writeup #384 today, which looks great btw!
I have a project with some namespaces where the last segment starts with a number like `db.migration.005-alter-table`. `suggested-aliases` offers `005-alter-table` as a possible alias, but that's not a valid suggestion since the clojure reader doesn't accept symbols that start with a number. When the nrepl respose is parsed by parseclj, it also correctly rejects that symbol, but that breaks cljr-slash for the entire project.
I put together this little change to remove such invalid symbols from the suggestions, and it's working well for me so far!

- [x] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (run `lein do clean, test`)
- [x] Code inlining with mranderson works and tests pass with inlined code (run `./build.sh install` -- takes a long time)
- [x] You've updated the changelog (if adding/changing user-visible functionality)

Thanks!

### CODE REVIEW

File: test/refactor_nrepl/ns/suggest_aliases_test.clj [hunk 1]
```diff
@@ -46,4 +46,7 @@
     "Removes redundant bits such as `clj-` and `.core`"
     'clj-a.b.c.core '{c     [clj-a.b.c.core]
                       b.c   [clj-a.b.c.core]
-                      a.b.c [clj-a.b.c.core]}))
+                      a.b.c [clj-a.b.c.core]}
+
+    "Removes fragments that start with a number, since they aren't valid symbols"
```
Comment (vemv) [inline test/refactor_nrepl/ns/suggest_aliases_test.clj:51]:
```suggestion
    "Removes fragments that start with a number (in this case: `9d`), since they aren't valid symbols"
```

**Review by vemv — APPROVED** [review]
Thanks much!

Kudos for finding the accurate cause of your problem.

Will be releasing this today as we merge this.

File: src/refactor_nrepl/ns/suggest_aliases.clj [hunk 2]
```diff
@@ -28,6 +28,9 @@
         fragments (map take-last
                        (range 1 (inc (count fragments)))
                        (repeat (distinct fragments)))
+        fragments (remove (fn [segments]
+                            (re-find #"^\d+" (first segments)))
+                          fragments)
```
Comment (vemv) [inline src/refactor_nrepl/ns/suggest_aliases.clj:33]:
```suggestion
```

Let's do the `remove` a few lines below:

```clj
(into {}
      (keep (fn [segments]
              (let [candidate (->> segments (string/join "."))]
                (when-not (invalid-symbol? candidate)
                  [(symbol candidate),
                   [namespace-name]]))))
      fragments)
```

where: 

```clj
(def invalid-symbol? (memoize (fn [x]
                                (try
                                  (edn/read-string x)
                                  false
                                  (catch Exception _
                                    true)))))
```

This way we gain genericity (my no. 1 concern - I don't want people to encounter similar problems in the future) and possibly performance.

Please make sure to unit-test `invalid-symbol?`

Thanks!

Comment (vemv) [reply src/refactor_nrepl/ns/suggest_aliases.clj:33]: (feel free to find a better name for `invalid-symbol?` - it accepts strings not symbols)

Comment (pdbrown) [reply src/refactor_nrepl/ns/suggest_aliases.clj]: Sure thing! I added a `symbol?` test too, since we probably shouldn't accept keywords etc either.

File: test/refactor_nrepl/ns/suggest_aliases_test.clj [hunk 3]
```diff
@@ -4,8 +4,8 @@
    [refactor-nrepl.ns.suggest-aliases :as sut]))
 
 (deftest test-like-ns-name?
-  (are [input expected] (= expected
-                           (sut/test-like-ns-name? input))
+  (are [input expected] (is (= expected
```
Comment (vemv) [inline test/refactor_nrepl/ns/suggest_aliases_test.clj:7]: the `is` is implied, please remove

File: test/refactor_nrepl/ns/suggest_aliases_test.clj [hunk 4]
```diff
@@ -21,6 +21,22 @@
     'unit.foo   true
     'foo.unit   true))
 
+(deftest readable-as-symbol?
+  (are [input expected] (is (= expected
```
Comment (vemv) [inline test/refactor_nrepl/ns/suggest_aliases_test.clj:25]: no `is`

File: test/refactor_nrepl/ns/suggest_aliases_test.clj [hunk 5]
```diff
@@ -46,4 +62,7 @@
     "Removes redundant bits such as `clj-` and `.core`"
     'clj-a.b.c.core '{c     [clj-a.b.c.core]
                       b.c   [clj-a.b.c.core]
-                      a.b.c [clj-a.b.c.core]}))
+                      a.b.c [clj-a.b.c.core]}
+
+    "Removes fragments that start with a number (in this case: `9d`), since they aren't valid symbols"
```
Comment (vemv) [inline test/refactor_nrepl/ns/suggest_aliases_test.clj:67]:
```suggestion
    "Removes fragments that start with a number (in this case: `9d`, `8b.c.9d`), since they aren't valid symbols"
```

File: test/refactor_nrepl/ns/suggest_aliases_test.clj [hunk 6]
```diff
@@ -21,6 +21,22 @@
...
     'foo.unit   true))
 
+(deftest readable-as-symbol?
+  (are [input expected] (is (= expected
+                               (sut/readable-as-symbol? input)))
+    "test"     true
+    "a.test"   true
+    "a-test"   true
+    "b.a-test" true
+    "ok then"  true
```
Comment (vemv) [inline test/refactor_nrepl/ns/suggest_aliases_test.clj:31]: while I don't expect `ok then` be produced irl, it is a good edge case to think of.

If it's at hand, the result should be `false` because a symbol with a space can be surprising / unusable client-side.

Then again I don't expect it to be a real problem, so feel free not to spend time in it

(in which case, remove this line so that we remain agnostic)

**Review by vemv — COMMENTED** [review]
> Sure thing! I added a symbol? test too, since we probably shouldn't accept keywords etc either.

Good thinking, thanks!

File: test/refactor_nrepl/ns/suggest_aliases_test.clj (hunk 4 above)
Comment (pdbrown) [reply test/refactor_nrepl/ns/suggest_aliases_test.clj:25]: Oh yes, that was a lazy hack to see the failing line `expected: (is (= false (sut/readable-as-symbol? "ok.one.more!")))` vs `expected: false`. I put in a proper fix: a plain `(readable-as-symbol? expected)` does the trick vs `(is (= true (...)))`.

File: test/refactor_nrepl/ns/suggest_aliases_test.clj (hunk 6 above)
Comment (pdbrown) [reply test/refactor_nrepl/ns/suggest_aliases_test.clj:31]: Good point, took it out.

File: test/refactor_nrepl/ns/suggest_aliases_test.clj (hunk 4 above)
Comment (vemv) [reply test/refactor_nrepl/ns/suggest_aliases_test.clj]: Yep, I understand!

That's why there's a `true` at tail position in this ns's `are`s (kondo and Eastwood won't complain - they understand the trick)

**Comment by vemv** [comment]
Thanks much and kudos for driving it to completion!

I'll release these today and you can count on it being available by tomorrow morning.

btw, I don't know if you were using the [suggest](https://github.com/clojure-emacs/clj-refactor.el/blob/1bd7b09976c05ebd1ff1ee3d5c882af303780b67/clj-refactor.el#L1945)  feature actively. I am, and it's been such a nice QoL improvement.

https://github.com/clojure-emacs/refactor-nrepl/issues/384 will be interesting as well, backing a revamped `cljr-slash`.

**Comment by vemv** [comment]
Available in clj-refactor (3.5.6; will be released within a couple hours) / refactor-nrepl (3.5.5; released)

**Comment by pdbrown** [comment]
Awesome, and thanks for all the quick reviews! Yeah, I just started using it and it's very helpful (which is how I noticed it wasn't working in my biggest project). I actually recently enabled clj-refactor again after not using it for a few years, and it's so great! I'd disabled it back in the day when I couldn't figure out how to prevent it from building ASTs, and perf was killing me due to my slow laptop + frequent repl restarts before I'd learned how to make things more reloadable.
The new suggest-libspecs op looks cool too! I'm interested in CLJC awareness and also smaller nREPL responses. I've noticed CIDER's `bdecode` can create a lot of garbage, so anything we do to shrink responses should help.

**Comment by vemv** [comment]
Awesome! That's great and motivating input to hear.

AST building can indeed be slow, it cannot scale indefinitely but that's just how `tools.analyzer` is.

Normally I work on 'modular' projects anyway so no codebase will be overly big.

We also possibly have a protocol that would use kondo / lsp instead. There's a lot to be gained with such hybrids.
